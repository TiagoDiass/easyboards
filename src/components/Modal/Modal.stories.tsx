import { ComponentMeta, ComponentStory } from '@storybook/react';
import Button from '../Button/Button';
import Modal from './Modal';
import { useModalState } from 'hooks';

export default {
  title: 'Modal',
  component: Modal,
  argTypes: {
    onClose: {
      action: 'closed'
    },

    isOpen: {
      type: 'symbol'
    },

    size: {
      options: ['xsmall', 'small', 'medium', 'large'],
      control: { type: 'radio' }
    }
  },
  args: {
    title: 'Modal title',
    size: 'medium'
  }
} as ComponentMeta<typeof Modal>;

export const Basic: ComponentStory<typeof Modal> = (args) => {
  const [isModalOpen, openModal, closeModal] = useModalState();

  return (
    <div>
      <h2 style={{ marginBottom: 24 }}>
        I recommend you to put your controls tab aside of the canvas <br />
        To do that, press <kbd>D</kbd>
      </h2>

      <Button color='success' onClick={openModal}>
        Open Modal
      </Button>

      <Modal {...args} isOpen={isModalOpen} onClose={closeModal}>
        <Modal.Content>
          <h2>Congratulations</h2>
          <p>You are inside a modal ðŸ˜‰</p>
        </Modal.Content>

        <Modal.Footer>
          <Button onClick={closeModal} color='danger'>
            Close modal
          </Button>
        </Modal.Footer>
      </Modal>
    </div>
  );
};
