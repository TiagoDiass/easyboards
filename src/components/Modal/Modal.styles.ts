import styled, { css } from 'styled-components';
import { ModalSizes } from './Modal';

export const Backdrop = styled.div`
  width: 100%;
  height: 100%;
  background: rgba(0, 0, 0, 0.3);
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const wrapperModifiers = {
  small: () => css`
    width: 500px;
    min-height: 420px;
  `,

  medium: () => css`
    width: 700px;
    min-height: 540px;
  `,

  large: () => css`
    width: 1100px;
    min-height: 680px;
  `
};

type WrapperProps = {
  size: ModalSizes;
};

export const Wrapper = styled.div<WrapperProps>`
  ${({ size, theme }) => css`
    box-shadow: 0 5px 16px rgba(0, 0, 0, 0.2);
    background: ${theme.colors.primary.accent1};
    color: ${theme.colors.secondary.accent2};
    position: relative;
    z-index: 10;
    border-radius: 10px;

    display: flex;
    flex-direction: column;

    ${!!size && wrapperModifiers[size]()}
  `}
`;

export const Header = styled.div`
  ${({ theme }) => css`
    border-bottom: 1px solid ${theme.colors.primary.accent2};

    display: flex;
    justify-content: center;
    align-items: center;
    padding: ${theme.spacings.xsmall} 0;
    position: relative;

    .title {
      font-size: ${theme.font.sizes.xlarge};
      color: ${theme.colors.secondary.accent2};
      font-weight: 700;
    }

    .close-button {
      position: absolute;
      right: 0.8rem;
      border-radius: 50%;
      width: 3.6rem;
      height: 3.6rem;
      padding: 0;
      display: flex;
      justify-content: center;
      align-items: center;

      // inner icon
      svg {
        width: 2rem;
      }

      // button inner content (<span />)
      span {
        display: none;
      }
    }
  `}
`;

export const ContentWrapper = styled.div`
  border: 1px solid red;
`;
