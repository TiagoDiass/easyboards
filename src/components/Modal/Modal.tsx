import * as S from './Modal.styles';
import { Close as CloseIcon } from '@styled-icons/evil';
import { Button } from 'components';
import ModalContent from './Elements/Content';
import ModalFooter, { ModalFooterProps } from './Elements/Footer';
import { DefaultTheme, StyledComponent } from 'styled-components';
import { useRef } from 'react';
import { useOutsideClick } from 'hooks';

export type ModalSizes = 'xsmall' | 'small' | 'medium' | 'large';

export type ModalProps = {
  title?: string;
  onClose?: () => void;
  size?: ModalSizes;
  isOpen?: boolean;
  children?: React.ReactNode;
};

interface CompoundedModal extends React.FunctionComponent<ModalProps> {
  /**
   * Compounded component that is used with the content of the Modal
   */
  Content: StyledComponent<'div', DefaultTheme>;

  /**
   * Compounded component that is used with the footer of the Modal
   */
  Footer: StyledComponent<'div', DefaultTheme, ModalFooterProps>;
}

/**
 * Modal component
 */
const Modal: CompoundedModal = ({
  title = 'Be careful',
  size = 'medium',
  isOpen = false,
  children,
  onClose
}: ModalProps) => {
  const modalContentWrapperRef = useRef(null);

  useOutsideClick(modalContentWrapperRef, onClose);

  return isOpen ? (
    <S.Backdrop>
      <S.Wrapper
        size={size}
        ref={modalContentWrapperRef}
        aria-label={`Modal with title "${title}"`}
      >
        <S.Header>
          <h3 className='title'>{title}</h3>

          <Button
            className='close-button'
            icon={<CloseIcon />}
            aria-label='Close modal'
            onClick={onClose}
          />
        </S.Header>

        {children}
      </S.Wrapper>
    </S.Backdrop>
  ) : null;
};

Modal.Content = ModalContent;
Modal.Footer = ModalFooter;

export default Modal;
