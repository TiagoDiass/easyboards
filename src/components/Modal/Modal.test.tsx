import Modal from './Modal';
import { renderWithTheme } from 'utils/test-utils';

xdescribe('Component: Modal', () => {
  it('should render correctly', () => {
    renderWithTheme(<Modal />);
  });
});
