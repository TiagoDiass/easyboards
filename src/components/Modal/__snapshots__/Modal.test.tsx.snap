// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Modal should render correctly 1`] = `
.c0 {
  width: 100%;
  height: 100%;
  background: rgba(0,0,0,0.3);
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-animation: jBcSpD 0.1s ease-in;
  animation: jBcSpD 0.1s ease-in;
}

.c1 {
  box-shadow: 0 5px 16px rgba(0,0,0,0.2);
  background: #FAFAFA;
  color: #333;
  position: relative;
  z-index: 10;
  border-radius: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 700px;
  height: 540px;
  -webkit-animation: gRMlCK 0.2s ease-in;
  animation: gRMlCK 0.2s ease-in;
}

.c2 {
  border-bottom: 1px solid #EAEAEA;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1.6rem 0;
  position: relative;
}

.c2 .title {
  font-size: 2.0rem;
  color: #333;
  font-weight: 700;
}

.c2 .close-button {
  position: absolute;
  right: 0.8rem;
  border-radius: 50%;
  width: 3.6rem;
  height: 3.6rem;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 .close-button svg {
  width: 2rem;
}

.c2 .close-button span {
  display: none;
}

.c4 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c3 {
  cursor: pointer;
  outline: none;
  border: 1px solid transparent;
  border-radius: 0.4rem;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #444;
  background-color: #FAFAFA;
  border-color: #EAEAEA;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

.c3:hover:not(:disabled) {
  background-color: #EAEAEA;
}

.c3:focus:not(:active) {
  box-shadow: 0 0 0 1px transparent,0 0 0 1px #EAEAEA;
}

.c3:disabled {
  opacity: 0.7;
  cursor: not-allowed;
}

.c3 svg {
  width: 1.6rem;
}

.c3 svg + span {
  margin-left: 0.8rem;
}

.c5 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  padding: 0 0.8rem;
  overflow-y: scroll;
}

.c5::-webkit-scrollbar {
  width: 4px;
}

.c5::-webkit-scrollbar-thumb {
  background: #333;
}

.c6 {
  border-top: 1px solid #EAEAEA;
  padding: 0.8rem;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <h3
        class="title"
      >
        Modal Title
      </h3>
      <button
        aria-disabled="false"
        aria-label="Close modal"
        class="c3 close-button"
        color="primary"
      >
        <svg
          aria-hidden="true"
          class="c4"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 50 50"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M37.304 11.282l1.414 1.414-26.022 26.02-1.414-1.413z"
          />
          <path
            d="M12.696 11.282l26.022 26.02-1.414 1.415-26.022-26.02z"
          />
        </svg>
        <span />
      </button>
    </div>
    <div
      class="c5"
    >
      <h1>
        My Content
      </h1>
    </div>
    <div
      class="c6"
    >
      <h1>
        Footer Content
      </h1>
    </div>
  </div>
</div>
`;
